### ISSUE TITLE
A property with generic protocol disables type checking when passing types instead of their instances

### ISSUE THREAD

**Issue by zmievsa** [microsoft/pyright#6788]
**Describe the bug**
When using a property with generics and protocols, somehow pyright disables all type checking when you pass a class instead of an instance of a class within a function.

**Code or Screenshots**
```python
from typing import Collection, Protocol, TypeVar, Generic
import ast

NodeT = TypeVar("NodeT", bound=ast.AST)


class MyProtocol(Protocol, Generic[NodeT]):
    @property
    def node_type(self) -> type[NodeT]:
        ...

    def __call__(self, node: NodeT) -> None:
        ...


class WrongImplementer:
    node_type = ast.ClassDef

    def __call__(self, boom: ast.Module, zoom: int, doom: str) -> None:
        return None


def hello(foo: Collection[MyProtocol]):
    pass


hello([WrongImplementer])  # No error
hello([WrongImplementer()])  # Error, as expected

```

Error text in the correct case:

```
Argument of type "list[WrongImplementer]" cannot be assigned to parameter "foo" of type "Collection[MyProtocol[Unknown]]" in function "hello"
  "WrongImplementer" is incompatible with protocol "MyProtocol[Unknown]"
    "__call__" is an incompatible type
      Type "(boom: Module, zoom: int, doom: str) -> None" cannot be assigned to type "(node: NodeT@MyProtocol) -> None"
        Parameter name mismatch: "node" versus "boom"
        Parameter 1: type "NodeT@MyProtocol" cannot be assigned to type "Module"
          "node_type" is incompatible with "Module"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

**VS Code extension or command-line**
VS Code Pylance: v2023.12.1

**Comment by erictraut** [comment]
I think pyright's behavior is correct here, as unintuitive as it may seem.

You've defined a generic protocol that takes a single type argument. In the signature for the `hello` function, you've omitted the type argument, so it is implicitly `Any`. 

A type that is compatible with your protocol must satisfy the following requirements:
1. It must allow read access to an attribute named `node_type` that produces a value of type `type[NodeT]` (which is specialized to `type[Any]` in the example).
2. It must supply a method `__call__` that accepts a single argument of type `NodeT`.

In the case that you reported as a false negative, you're passing a value of type `Collection[type[WrongImplementer]]` rather than `Collection[WrongImplementer]`, so we have to determine whether a value of type `type[WrongImplementer]` satisfies the protocol's requirements. It turns out that it does meet these requirements.
* It provides a class variable named `node_type` which is compatible with type `type[Any]`.
* A value of type `type[WrongImplementer]` is a subtype of the class `type`, which provides a `__call__` method defined as follows in the typeshed definition of `type`.

```python
class type:
    ...
    def __call__(self, *args: Any, **kwds: Any) -> Any: ...
```

This signature is compatible with the `__call__` defined in the protocol.

So, I think pyright is doing the right thing here, and there is no bug to fix.

**Comment by zmievsa** [comment]
Wow. This is crazy. Thank you for a thorough explanation?
